### PR TITLE
util, rxm: Updates to util_wait code

### PR DESCRIPTION
**Pull request by a-ilango**
- prov/util: Augment util wait code to wait on additional file descriptors
- prov/rxm: Use ofi_cq_sread calls, add msg_cq fd to util_wait fd list

### CODE REVIEW

File: include/fi_util.h [hunk 1]
```diff
@@ -542,10 +542,23 @@ struct util_wait_fd {
 	struct util_wait	util_wait;
 	struct fd_signal	signal;
 	fi_epoll_t		epoll_fd;
+	struct dlist_entry	fd_list;
+};
+
+typedef int (*ofi_wait_fd_try_func)(const void *arg);
```
Comment (shefty) [inline include/fi_util.h]: Should the parameter be const?

File: include/fi_util.h [hunk 2]
```diff
@@ -542,10 +542,23 @@ struct util_wait_fd {
 	struct util_wait	util_wait;
 	struct fd_signal	signal;
 	fi_epoll_t		epoll_fd;
+	struct dlist_entry	fd_list;
```
Comment (shefty) [inline include/fi_util.h]: Locking is needed around fd_list.

File: prov/util/src/util_wait.c [hunk 3]
```diff
@@ -212,13 +275,20 @@ static int util_wait_fd_control(struct fid *fid, int command, void *arg)
...
 	wait = container_of(fid, struct util_wait_fd, util_wait.wait_fid.fid);
 	ret = fi_wait_cleanup(&wait->util_wait);
 	if (ret)
 		return ret;
 
+	while (!dlist_empty(&wait->fd_list)) {
+		dlist_pop_front(&wait->fd_list, struct ofi_wait_fd_entry,
+				fd_entry, entry);
+		free(fd_entry);
+	}
```
Comment (shefty) [inline prov/util/src/util_wait.c]: A non-empty fd_list seems like an error.  Should we just fail the close call?

File: include/fi_util.h (hunk 1 above)
Comment (a-ilango) [inline include/fi_util.h]: The function is not supposed to modify arg value. So added a const there.

File: include/fi_util.h (hunk 2 above)
Comment (a-ilango) [inline include/fi_util.h]: Yes. Will add a lock to access fd_list.

File: include/fi_util.h (hunk 1 above)
Comment (shefty) [inline include/fi_util.h]: I think the try call implies the argument will be modified in some way.  An easy way to check this is to remove all casts from the try callback and declare any local variable as const.  Then see if a cast ends up being necessary to remove the const.

File: prov/util/src/util_wait.c (hunk 3 above)
Comment (a-ilango) [inline prov/util/src/util_wait.c]: Ideally, when the endpoint is closed, it should remove the fd entry from fd_list. This code takes care of the case when an endpoint doesn't do it. Perhaps this code can be removed as it's an error if the endpoint doesn't delete the fd from the list. May be we should use a reference counter on the util_wait object?

**Comment by shefty** [comment]
I thought about adding a reference count, but I think simply checking that the list is empty should be sufficient.  We can make that an assert, as it should never happen.

File: prov/util/src/util_wait.c [hunk 4]
```diff
@@ -142,6 +142,62 @@ int fi_wait_init(struct util_fabric *fabric, struct fi_wait_attr *attr,
...
+
+int ofi_wait_fd_add(struct util_wait *wait, int fd, ofi_wait_fd_try_func try,
+		    void *arg, void *context)
+{
+	struct ofi_wait_fd_entry *fd_entry;
+	struct util_wait_fd *wait_fd = container_of(wait, struct util_wait_fd,
+						    util_wait);
+	int ret;
+
+	ret = fi_epoll_add(wait_fd->epoll_fd, fd, context);
```
Comment (shefty) [inline prov/util/src/util_wait.c]: We need to check for duplicate insertions

**Comment by shefty** [comment]
Added do no merge, since I thought you mentioned this was post 1.5.  If this is for 1.5, please remove that tag

File: prov/util/src/util_wait.c [hunk 5]
```diff
@@ -142,6 +142,68 @@ int fi_wait_init(struct util_fabric *fabric, struct fi_wait_attr *attr,
...
+
+	fastlock_acquire(&wait_fd->lock);
+	entry = dlist_remove_first_match(&wait_fd->fd_list, ofi_wait_fd_match,
+					 &fd);
+	fastlock_release(&wait_fd->lock);
+	if (!entry) {
+		FI_WARN(wait->prov, FI_LOG_FABRIC,
+			"Given fd not found in wait list\n");
+		return -FI_EINVAL;
+	}
```
Comment (shefty) [inline prov/util/src/util_wait.c]: I think an fd might be added to the same wait set twice, so I would convert the FI_WARN into FI_INFO

File: prov/util/src/util_wait.c [hunk 6]
```diff
@@ -142,6 +142,68 @@ int fi_wait_init(struct util_fabric *fabric, struct fi_wait_attr *attr,
...
+						    util_wait);
+	int ret = 0;
+
+	fastlock_acquire(&wait_fd->lock);
+	if (dlist_find_first_match(&wait_fd->fd_list, ofi_wait_fd_match, &fd))
+		goto out;
+
+	ret = fi_epoll_add(wait_fd->epoll_fd, fd, context);
+	if (ret)
+		goto out;
```
Comment (shefty) [inline prov/util/src/util_wait.c]: Add an FI_WARN on error here.  We should be better about adding error messages on error paths to assist with future debugging efforts.

**Comment by shefty** [comment]
This is failing on windows debug build.  Re-starting the test shows the same failure.  Please see the appveyor results.

**Comment by a-ilango** [comment]
Appveyor build ran twice after updating the PR. I didn't see the failure again.

File: prov/util/src/util_wait.c [hunk 7]
```diff
@@ -142,6 +142,73 @@ int fi_wait_init(struct util_fabric *fabric, struct fi_wait_attr *attr,
 	return 0;
 }
 
+int ofi_wait_fd_match(struct dlist_entry *item, const void *arg)
```
Comment (gladkovdmitry17) [inline prov/util/src/util_wait.c]: Should it be marked as `static`?

Comment (a-ilango) [inline prov/util/src/util_wait.c]: Yes, that should be static. I've updated the code.

**Pull request merged**